src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+                }
+            }
+
+            return root;
+        }
+
+        private static bool IsList(ITypeSymbol typeSymbol, SemanticModel semanticModel)
+            => Equals(typeSymbol.OriginalDefinition, semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName));
+
+        private abstract class AbstractConverter
+        {
+            protected static QueryExpressionSyntax CreateQueryExpression(ForEachStatementSyntax forEachStatement, ImmutableArray<SyntaxNode> convertingNodes, ExpressionSyntax selectExpression)
+                => SyntaxFactory.QueryExpression(
+                    CreateFromClause(forEachStatement),
+                    SyntaxFactory.QueryBody(
+                        SyntaxFactory.List(convertingNodes.Select(qc => CreateQueryClause(qc).WithoutTrivia())),
+                        // The current coverage of foreach statements to support does not need to use query continuations.
+                        SyntaxFactory.SelectClause(selectExpression).WithoutTrivia(), continuation: null).WithoutTrivia()).WithAdditionalAnnotations(Formatter.Annotation);
+
+            private static QueryClauseSyntax CreateQueryClause(SyntaxNode node)
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.VariableDeclarator:
+                        var variable = (VariableDeclaratorSyntax)node;
+                        return SyntaxFactory.LetClause(variable.Identifier, variable.Initializer.Value);
+                    case SyntaxKind.ForEachStatement:
+                        return CreateFromClause((ForEachStatementSyntax)node);
+                    case SyntaxKind.IfStatement:
+                        return SyntaxFactory.WhereClause(((IfStatementSyntax)node).Condition);
+                }
+
+                throw new ArgumentException(node.Kind().ToString());
+            }
+
+            private static FromClauseSyntax CreateFromClause(ForEachStatementSyntax forEachStatement)
+                => SyntaxFactory.FromClause(
+                    forEachStatement.Type is IdentifierNameSyntax identifierName && identifierName.Identifier.ValueText == "var" ? null : forEachStatement.Type,
+                    forEachStatement.Identifier,
+                    forEachStatement.Expression);
+        }
+
+        private sealed class YieldReturnConverter : AbstractConverter, IConverter
+        {
+            private ForEachStatementSyntax _forEachStatement;
+            private ImmutableArray<SyntaxNode> _convertingNodes;
+            private YieldStatementSyntax _yieldStatement;
+            private SyntaxNode _nodeToDelete;
+
+            public YieldReturnConverter(
+                ForEachStatementSyntax forEachStatement,
+                ImmutableArray<SyntaxNode> convertingNodes,
+                YieldStatementSyntax yieldStatement,
+                SyntaxNode nodeToDelete = null)
```

Comment:
making this optional seems strange.  In general, in private code the IDE doesn't use optional as it means callers sometimes forget data they need to pass.